### PR TITLE
Get Halium 10 recovery to install updates

### PR DESCRIPTION
**Pull request by NotKit**
Tested on Xiaomi Poco M3 (citrus), Xiaomi Redmi Note 9 (merlin) and Xiaomi Redmi Note 9 Pro/9S (miatoll) devices.

### CODE REVIEW

**Comment by amartinz** [comment]
Verified to work with SHIFT6mq (axolotl) on H10, which uses dynamic partitions and A/B.

**Comment by amartinz** [comment]
Though this renders the recovery itself kind of unusable.

eg the userspace fastboot (fastbootd) mode is broken now and i can not flash super partitions anymore.

```
0 ✓ amartinz axolotl [android-10] $> fastboot flash system ota_out/system.img 
Resizing 'system_a'                                OKAY [  0.005s]
Sending sparse 'system_a' 1/5 (486165 KB)          OKAY [  2.297s]
Writing 'system_a'                                 FAILED (remote: 'No such file or directory')
```

**Comment by amartinz** [comment]
Couldn't we always mount /system while in recovery and switch over to `/mnt/system` for the script?

https://github.com/ubports/halium_bootable_recovery/blob/halium-10.0/recovery.cpp#L705

**Comment by Flohack74** [comment]
Wasnt the system partition supposed to be in /system_root anyways?

**Comment by amartinz** [comment]
> Wasnt the system partition supposed to be in /system_root anyways?

Only for System-As-Root devices, which are deprecated by Android.
For H10 dynamic partitions is the new, cool kid on the block.

**Comment by Flohack74** [comment]
Ok I see so I will take a look tonight if this is ok or not ;)

File: recovery.cpp [hunk 1]
```diff
@@ -79,7 +79,7 @@ static constexpr const char* LAST_KMSG_FILE = "/cache/recovery/last_kmsg";
 static constexpr const char* LAST_LOG_FILE = "/cache/recovery/last_log";
 static constexpr const char* LOCALE_FILE = "/cache/recovery/last_locale";
 
-static constexpr const char* CACHE_ROOT = "/cache";
+/*static constexpr const char* CACHE_ROOT = "/cache";*/
```
Comment (Flohack74) [inline recovery.cpp:82]: Do not leave commented code, simply delete it

File: recovery.cpp [hunk 2]
```diff
@@ -172,7 +172,8 @@ static void finish_recovery() {
     if (ensure_path_mounted(COMMAND_FILE) != 0 || (unlink(COMMAND_FILE) && errno != ENOENT)) {
       LOG(WARNING) << "Can't unlink " << COMMAND_FILE;
     }
-    ensure_path_unmounted(CACHE_ROOT);
+    // Keep /cache mounted, needed by UBPorts Installer
```
Comment (Flohack74) [inline recovery.cpp:175]: Same

File: ubports/system-image-upgrader [hunk 3]
```diff
@@ -452,6 +453,7 @@ do
                         umount $SYSTEM_PARTITION || true
                         umount $SYSTEM_MOUNTPOINT || true
                         umount /system_root || true
```
Comment (Flohack74) [inline ubports/system-image-upgrader:455]: Do we need /system_root here now? Shouldnt that be deprecated as @amartinz stated?

File: otautil/roots.cpp [hunk 4]
```diff
@@ -90,10 +90,16 @@ void load_volume_table() {
...
 
+  // Map logical partitions so there is a device node to put into fstab
+  setup_install_mounts();
+
   // Create a boring /etc/fstab so tools like Busybox work
   FILE* file = fopen("/etc/fstab", "w");
   if (file) {
     for (auto& entry : fake_fstab) {
+      if (entry.fs_mgr_flags.logical && !fs_mgr_update_logical_partition(&entry)) {
+        LOG(ERROR) << "Failed to find block device for partition" << entry.blk_device;
```
Comment (Flohack74) [inline otautil/roots.cpp:101]: What happens in this case? Will an eventual update crash or get stuck? We should not leave the device in a half-installed state if possible

**Review by Flohack74 — CHANGES_REQUESTED** [review]

File: ubports/system-image-upgrader (hunk 3 above)
Comment (amartinz) [reply ubports/system-image-upgrader:455]: well, devices upgrading to Android 10 can still have it while devices launching with Android 10 usually do not have it.

so we have to keep this to not break devices, which already use it

**Pull request closed without merging**